txnkv/transaction/2pc.go

```diff
@@ -1265,19 +1343,119 @@ func keepAlive(
 					}
 					return
 				}
-				continue
+			} else {
+				keepFail = 0
+				metrics.TxnHeartBeatHistogramOK.Observe(time.Since(startTime).Seconds())
+			}
+
+			// broadcast to all stores
+			if isPipelinedTxn {
+				broadcastToAllStores(
+					c.txn,
+					c.store,
+					retry.NewBackofferWithVars(
+						context.Background(),
+						broadcastMaxBackoff,
+						c.txn.vars,
+					),
+					&kvrpcpb.TxnStatus{
+						StartTs:     c.startTS,
+						MinCommitTs: c.minCommitTSMgr.get(),
+						CommitTs:    0,
+						RolledBack:  false,
+						IsCompleted: false,
+					},
+					c.resourceGroupName,
+					c.resourceGroupTag,
+				)
 			}
-			keepFail = 0
-			metrics.TxnHeartBeatHistogramOK.Observe(time.Since(startTime).Seconds())
 		}
 	}
 }
 
-func sendTxnHeartBeat(bo *retry.Backoffer, store kvstore, primary []byte, startTS, ttl uint64) (newTTL uint64, stopHeartBeat bool, err error) {
+const broadcastRpcTimeout = time.Second * 5
+const broadcastMaxConcurrency = 10
+
+// broadcastToAllStores asynchronously broadcasts the transaction status to all stores
+// errors are ignored.
+func broadcastToAllStores(
+	txn *KVTxn,
+	store kvstore,
+	bo *retry.Backoffer,
+	status *kvrpcpb.TxnStatus,
+	resourceGroupName string,
+	resourceGroupTag []byte,
+) {
+	broadcastFunc := func() {
+		stores := store.GetRegionCache().GetStoresByType(tikvrpc.TiKV)
+		req := tikvrpc.NewRequest(
+			tikvrpc.CmdBroadcastTxnStatus, &kvrpcpb.BroadcastTxnStatusRequest{
+				TxnStatus: []*kvrpcpb.TxnStatus{status},
+			},
+		)
+		req.Context.ClusterId = store.GetClusterID()
+		req.Context.ResourceControlContext = &kvrpcpb.ResourceControlContext{
+			ResourceGroupName: resourceGroupName,
+		}
+		req.Context.ResourceGroupTag = resourceGroupTag
+
+		var wg sync.WaitGroup
+		concurrency := min(broadcastMaxConcurrency, len(stores))
+		taskChan := make(chan *locate.Store, concurrency)
+
+		for i := 0; i < concurrency; i++ {
+			wg.Add(1)
+			if err := txn.spawnWithStorePool(func() {
+				defer wg.Done()
+				for s := range taskChan {
+					_, err := store.GetTiKVClient().SendRequest(
+						bo.GetCtx(),
+						s.GetAddr(),
+						req,
+						broadcastRpcTimeout,
+					)
+					if err != nil {
+						logutil.Logger(store.Ctx()).Info(
+							"broadcast txn status failed",
+							zap.Uint64("storeID", s.StoreID()),
+							zap.String("storeAddr", s.GetAddr()),
+							zap.Stringer("status", status),
+							zap.Error(err),
+						)
+					}
+				}
+			}); err != nil {
+				wg.Done() // Ensure wg is decremented if spawning fails
+				logutil.Logger(store.Ctx()).Error("failed to spawn worker goroutine", zap.Error(err))
+			}
+		}
+
+		for _, s := range stores {
+			taskChan <- s
+		}
+
+		close(taskChan)
+		wg.Wait()
+	}
+
+	if err := txn.spawnWithStorePool(broadcastFunc); err != nil {
```

Comment:
Since we've already used a goroutine pool, how about spawning broadcast tasks like the following:
```go
	stores := store.GetRegionCache().GetStoresByType(tikvrpc.TiKV)
	rateLimit := make(chan struct{}, min(broadcastMaxConcurrency, len(stores)))
	for _, s := range stores {
		rateLimit <- struct{}{}
		target := s
		err := txn.spawnWithStorePool(func() {
			defer func() {
				<-rateLimit
			}()
			// send request to target
		})
		if err != nil {
			<-rateLimit
		}
	}
```